streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```diff
@@ -102,11 +120,29 @@ public void makeReady(final Map<String, InternalTopicConfig> topics) {
             final Set<NewTopic> newTopics = new HashSet<>();
 
             for (final InternalTopicConfig internalTopicConfig : topicsToBeCreated) {
-                final Properties topicProperties = internalTopicConfig.toProperties(windowChangeLogAdditionalRetention);
-                final Map<String, String> topicConfig = new HashMap<>(defaultTopicConfigs);
-                for (final String key : topicProperties.stringPropertyNames()) {
-                    topicConfig.put(key, topicProperties.getProperty(key));
+                // internal topic config overridden rule: library overrides < global config overrides < per-topic config overrides
+                final Map<String, String> topicConfig = new HashMap<>();
+
+                switch (internalTopicConfig.type()) {
```

Comment:
This and other code, i.e, `if(topicType == ..)` makes me think we should maybe have an interface and three implementations of this. Then we can remove the `if` and `switch` around the type and just ask the impl to do whatever it needs to

Reply:
Not sure I understand this statement, could you elaborate a bit more? Currently this `switch` pattern is only called once so I'm not sure if it is worthwhile to refactor it as an interface, or maybe I completely got it wrong since you mention `this and other code` (while I cannot think of `other` code?).

Reply:
I think what Damian meant is, that we need to pass in the internal topic type in the constructor -- instead of that, we can just instantiate different types directly.

Reply:
I see your point now. The reason I wanted to pass in the `topicType` at constructor is that, the `setRetentionMs()` can be guarded earlier like we did today: if we do not know the `topicType` we will blindly accept the retention ms.